File: crates/muse-local-api/src/dto/preferences.rs
//! 应用偏好接口的请求、响应与错误 DTO。

use serde::{Deserialize, Serialize};

/// 外观偏好更新请求。
#[derive(Deserialize)]
pub(crate) struct AppearancePreferencesUpdateRequest {
    pub(crate) background_blur: u8,
    pub(crate) background_opacity: f64,
    pub(crate) motion_level: muse_core::app::preferences::MotionLevel,
}

/// 外观偏好与配置诊断响应。
#[derive(Serialize)]
pub(crate) struct AppearancePreferencesResponse {
    schema_version: u32,
    appearance: muse_core::app::preferences::AppearancePreferences,
    diagnostics: Vec<muse_core::app::preferences::ConfigDiagnostic>,
}

/// 用户配置变更失败响应。
#[derive(Serialize)]
pub(crate) struct ConfigMutationErrorResponse {
    pub(crate) error: String,
    pub(crate) code: String,
    pub(crate) field_path: String,
    pub(crate) message: String,
}

impl From<muse_core::app::preferences::MuseConfigSnapshot> for AppearancePreferencesResponse {
    fn from(snapshot: muse_core::app::preferences::MuseConfigSnapshot) -> Self {
        Self {
            schema_version: snapshot.config.schema_version,
            appearance: snapshot.config.appearance,
            diagnostics: snapshot.diagnostics,
        }
    }
}

File: src/api/preferences.ts
import type { AppearanceBackgroundTheme, AppearanceTheme } from '@/types';
import { apiFetch, readJson } from './client';

export type AppearanceMotionLevel = 'full' | 'reduced' | 'none';

export interface ConfigDiagnostic {
  code: string;
  field_path: string;
  message: string;
}

export interface AppearancePreferencesResponse {
  schema_version: number;
  appearance: {
    theme: AppearanceTheme;
    background_theme: AppearanceBackgroundTheme;
    language: string;
    background_blur: number;
    background_opacity: number;
    motion_level: AppearanceMotionLevel;
  };
  diagnostics: ConfigDiagnostic[];
}

export interface AppearancePreferencesUpdate {
  theme: AppearanceTheme;
  background_theme: AppearanceBackgroundTheme;
  background_blur: number;
  background_opacity: number;
  motion_level: AppearanceMotionLevel;
}

/** 读取由用户级 config.toml 持久化的外观偏好。 */
export async function fetchAppearancePreferences(): Promise<AppearancePreferencesResponse> {
  return readJson<AppearancePreferencesResponse>(
    await apiFetch('/api/preferences/appearance')
  );
}

/** 只更新界面当前可编辑的外观字段，后端保留 TOML 中的其他字段和注释。 */
export async function saveAppearancePreferences(
  update: AppearancePreferencesUpdate
): Promise<AppearancePreferencesResponse> {
  return readJson<AppearancePreferencesResponse>(
    await apiFetch('/api/preferences/appearance', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(update)
    })
  );
}
